Sources/Disposable.swift

```diff
@@ -40,68 +40,81 @@ extension UnsafeAtomicState where State == DisposableState {
 
 /// A type-erased disposable that forwards operations to an underlying disposable.
 public final class AnyDisposable: Disposable {
-	private let disposable: Disposable
-
-	public var isDisposed: Bool {
-		return disposable.isDisposed
-	}
-
-	public init(_ disposable: Disposable) {
-		self.disposable = disposable
+	private enum Backing {
+		case wrapping(Disposable)
+
+		// We need to ability to release the closure. The apparent way would be packing an
+		// optional closure reference. However, the way Swift packs multi- payload enum
+		// discriminator via both extra inhabitants and unused bits may result in
+		// corrupted reads even after masking. So a constant pointer to an unsafe
+		// allocation is used instead.
+		case action(UnsafeAtomicState<DisposableState>, UnsafeMutablePointer<() -> Void>?)
 	}
 
-	public func dispose() {
-		disposable.dispose()
-	}
-}
-
-/// A disposable that only flips `isDisposed` upon disposal, and performs no other
-/// work.
-public final class SimpleDisposable: Disposable {
-	private var state = UnsafeAtomicState(DisposableState.active)
+	private let base: Backing
```

Comment:
I don't care for this implementation. I find this "multiple implementation via a backing enum" approach to be difficult to reason about.

If we kept `SimpleDisposable` and `ActionDisposable`, but made them `private` and added `init`s that exposed the functionality publicly, then I think I'd be onboard.